// jira-precommit-hook/lib/cli-commands/jokes.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _config = require('../config');

var _config2 = _interopRequireDefault(_config);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var command = 'jokes';

exports.default = {
  command: command,
  execute: function execute(_ref) {
    var _this = this;

    var decision = _ref.decision;
    return (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee() {
      return _regenerator2.default.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              if (decision === 'enable') {
                _config2.default.set('jokes', true);
                console.log('Jokes enabled!');
              } else if (decision === 'disable') {
                _config2.default.set('jokes', false);
                console.log('Jokes disabled!');
              } else if (decision === 'explicit') {
                _config2.default.set('explicit', true);
                console.log('Profane joke filter disabled!');
              } else if (decision === 'clean') {
                _config2.default.set('explicit', false);
                console.log('Profane joke filter enabled!');
              }

              return _context.abrupt('return', 0);

            case 2:
            case 'end':
              return _context.stop();
          }
        }
      }, _callee, _this);
    }))();
  },
  register: function register(subparsers) {
    var joke = subparsers.addParser(command, {
      addHelp: true,
      help: command + ' help'
    });

    joke.addArgument('decision', {
      choices: ['enable', 'disable', 'explicit', 'clean']
    });
  }
};
module.exports = exports['default'];
